crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -152,57 +161,308 @@ impl RawModuleResolutionSettings {
+/// Represents a single `.pth` file in a `site-packages` directory.
+/// One or more lines in a `.pth` file may be a (relative or absolute)
+/// path that represents an editable installation of a package.
+struct PthFile<'db> {
+    db: &'db dyn Db,
+    path: SystemPathBuf,
+    contents: SourceText,
+    site_packages: &'db SystemPath,
+}
+
+impl<'db> PthFile<'db> {
+    fn new(
+        db: &'db dyn Db,
+        pth_path: SystemPathBuf,
+        file: File,
+        site_packages: &'db SystemPath,
+    ) -> Self {
+        Self {
+            db,
+            path: pth_path,
+            contents: source_text(db.upcast(), file),
+            site_packages,
+        }
+    }
+
+    /// Yield paths in this `.pth` file that appear to represent editable installations,
+    /// and should therefore be added as module-resolution search paths.
+    fn iter_editable_installations<'a>(
+        &'a self,
+    ) -> impl Iterator<Item = ModuleResolutionPathBuf> + 'db
+    where
+        'a: 'db,
+    {
+        // Empty lines or lines starting with '#' are ignored by the Python interpreter.
+        // Lines that start with "import " or "import\t" do not represent editable installs at all;
+        // instead, these are files that are executed by Python at startup.
+        // https://docs.python.org/3/library/site.html#module-site
+        self.contents.lines().filter_map(|line| {
+            let line = line.trim();
+            if line.is_empty()
+                || line.starts_with('#')
+                || line.starts_with("import ")
+                || line.starts_with("import\t")
+            {
+                return None;
+            }
+            let possible_editable_install = SystemPath::absolute(
+                self.site_packages.join(line),
+                self.db.system().current_directory(),
+            );
+
+            ModuleResolutionPathBuf::editable_installation_root(self.db, possible_editable_install)
+        })
+    }
+}
+
+/// Iterator that yields a [`PthFile`] instance for every `.pth` file
+/// found in a given `site-packages` directory.
+struct PthFileIterator<'db> {
+    db: &'db dyn Db,
+    directory_iterator: Box<dyn Iterator<Item = std::io::Result<DirectoryEntry>> + 'db>,
+    site_packages: &'db SystemPath,
+}
+
+impl<'db> PthFileIterator<'db> {
+    fn new(db: &'db dyn Db, site_packages: &'db SystemPath) -> std::io::Result<Self> {
+        // TODO: The salsa cache should be invalidated every time a `.pth` file is added to `site-packages`,
+        // as well as every time an existing one is removed.
+        Ok(Self {
+            db,
+            directory_iterator: db.system().read_directory(site_packages)?,
+            site_packages,
+        })
     }
 }
 
+impl<'db> Iterator for PthFileIterator<'db> {
+    type Item = PthFile<'db>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let entry_result = self.directory_iterator.next()?;
+            let Ok(entry) = entry_result else {
```

Comment:
I think Charlie's implementation also has an upper bound on the `pth` file sizes. Should we add this too?

Reply:
I saw that. I believe it's copied directly from pyright's logic here: https://github.com/microsoft/pyright/blob/042d2ea58246f0e4a8e1597a540878badfae85b7/packages/pyright-internal/src/analyzer/pythonPathUtils.ts#L193-L206

I don't much like it: it seems pretty ad-hoc to me. There's nothing in the Python documentation that indicates that it will stop adding entries from `pth` files when they exceed a certain size. If you do end up with a massive `pth` file in `site-packages`, I don't think it's Ruff's fault if that creates a startup penalty for Ruff: it's probably the package the user installed, or the build backend that package is using, that's doing something wrong there.